### PR TITLE
added recalculatePartitions delay for node restarting

### PR DESCRIPTION
**Pull request by YevhenBondarenko**
## Pull Request description

If node removed then better to wait some time instead of immediately partitions recalculation. Its helpful for rollout restart.

## General checklist

- [ ] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [ ] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [ ] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [ ] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [ ] Description contains human-readable scope of changes.
- [ ] Description contains brief notes about what needs to be added to the documentation.
- [ ] No merge conflicts, commented blocks of code, code formatting issues.
- [ ] Changes are backward compatible or upgrade script is provided.
- [ ] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.
  
## Front-End feature checklist

- [ ] Screenshots with affected component(s) are added. The best option is to provide 2 screens: before and after changes;
- [ ] If you change the widget or other API, ensure it is backward-compatible or upgrade script is present.
- [ ] Ensure new API is documented [here](https://github.com/thingsboard/thingsboard-ui-help)

## Back-End feature checklist

- [ ] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [ ] If new dependency was added: the dependency tree is checked for conflicts.
- [ ] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [ ] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.

### CODE REVIEW

File: common/queue/src/main/java/org/thingsboard/server/queue/discovery/ZkDiscoveryService.java [hunk 1]
```diff
@@ -290,8 +297,30 @@ public void childEvent(CuratorFramework curatorFramework, PathChildrenCacheEvent
         log.debug("Processing [{}] event for [{}]", pathChildrenCacheEvent.getType(), instance.getServiceId());
         switch (pathChildrenCacheEvent.getType()) {
             case CHILD_ADDED:
+                ScheduledFuture<?> task = delayedTasks.remove(instance.getServiceId());
```
Comment (ashvayka) [inline common/queue/src/main/java/org/thingsboard/server/queue/discovery/ZkDiscoveryService.java:300]: I am not sure that ```childEvent``` is executed synchronously. So this may cause corner cases if two events for the same server arrive simultaneously.

File: common/queue/src/main/java/org/thingsboard/server/queue/discovery/ZkDiscoveryService.java [hunk 2]
```diff
@@ -290,8 +297,30 @@ public void childEvent(CuratorFramework curatorFramework, PathChildrenCacheEvent
...
+                if (task != null) {
```
Comment (ashvayka) [inline common/queue/src/main/java/org/thingsboard/server/queue/discovery/ZkDiscoveryService.java:301]: NO TESTS - NO MERGE

File: common/queue/src/main/java/org/thingsboard/server/queue/discovery/ZkDiscoveryService.java [hunk 3]
```diff
@@ -290,8 +297,32 @@ public void childEvent(CuratorFramework curatorFramework, PathChildrenCacheEvent
...
+                    log.debug("[{}] Going to recalculate partitions due to adding new node [{}].",
+                            instance.getServiceId(), instance.getServiceTypesList());
+                    recalculatePartitions();
+                }
+                break;
             case CHILD_REMOVED:
-                recalculatePartitions();
+                ScheduledFuture<?> future = zkExecutorService.schedule(() -> {
+                    log.debug("[{}] Going to recalculate partitions due to removed node [{}]",
+                            instance.getServiceId(), instance.getServiceTypesList());
```
Comment (smatvienko-tb) [inline common/queue/src/main/java/org/thingsboard/server/queue/discovery/ZkDiscoveryService.java:319]: Please, hold primitives in runnable, not instances. If it is small and immutable - it's ok as is

File: common/queue/src/main/java/org/thingsboard/server/queue/discovery/ZkDiscoveryService.java [hunk 4]
```diff
@@ -303,6 +334,7 @@ public void childEvent(CuratorFramework curatorFramework, PathChildrenCacheEvent
      * Synchronized to ensure that other servers info is up to date
      * */
     synchronized void recalculatePartitions() {
+        delayedTasks.clear();
```
Comment (smatvienko-tb) [inline common/queue/src/main/java/org/thingsboard/server/queue/discovery/ZkDiscoveryService.java:337]: good point to clean as repartition started, but futures from map not canceled and sits in scheduler

File: common/queue/src/test/java/org/thingsboard/server/queue/discovery/ZkDiscoveryServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        ReflectionTestUtils.setField(zkDiscoveryService, "cache", cache);
+        ReflectionTestUtils.setField(zkDiscoveryService, "nodePath", "/thingsboard/nodes/0000000010");
+        ReflectionTestUtils.setField(zkDiscoveryService, "zkExecutorService", zkExecutorService);
+        ReflectionTestUtils.setField(zkDiscoveryService, "recalculateDelay", 1000L);
+        ReflectionTestUtils.setField(zkDiscoveryService, "zkDir", "/thingsboard");
+    }
+
+    @Test
+    public void restartNodeTest() throws Exception {
+        var currentInfo = TransportProtos.ServiceInfo.newBuilder().setServiceId("currentId").build();
```
Comment (smatvienko-tb) [inline common/queue/src/test/java/org/thingsboard/server/queue/discovery/ZkDiscoveryServiceTest.java:89]: Please, use real-name services as examples. like 'tb-rule-engine-0', etc

File: common/queue/src/test/java/org/thingsboard/server/queue/discovery/ZkDiscoveryServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+        verify(partitionService, never()).recalculatePartitions(eq(currentInfo), any());
+
+        Thread.sleep(2000);
+
+        verify(partitionService, never()).recalculatePartitions(eq(currentInfo), any());
+
+        assertTrue(zkDiscoveryService.delayedTasks.isEmpty());
+
+        //Restart not in time
```
Comment (smatvienko-tb) [inline common/queue/src/test/java/org/thingsboard/server/queue/discovery/ZkDiscoveryServiceTest.java:122]: Please, split this into 3 or more tests, and remove empty lines. I can not open the whole test on Full HD laptop :)

File: common/queue/src/test/java/org/thingsboard/server/queue/discovery/ZkDiscoveryServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    public void restartNodeTest() throws Exception {
+        var currentInfo = TransportProtos.ServiceInfo.newBuilder().setServiceId("currentId").build();
+        var currentData = new ChildData("/thingsboard/nodes/0000000010", null, currentInfo.toByteArray());
+        var childInfo = TransportProtos.ServiceInfo.newBuilder().setServiceId("childId").build();
+        var childData = new ChildData("/thingsboard/nodes/0000000020", null, childInfo.toByteArray());
+
+        when(serviceInfoProvider.getServiceInfo()).thenReturn(currentInfo);
+        List<ChildData> dataList = new ArrayList<>();
+        dataList.add(currentData);
+        when(cache.getCurrentData()).thenReturn(dataList);
```
Comment (smatvienko-tb) [inline common/queue/src/test/java/org/thingsboard/server/queue/discovery/ZkDiscoveryServiceTest.java:97]: inline List.of(dataList)

File: common/queue/src/test/java/org/thingsboard/server/queue/discovery/ZkDiscoveryServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+        verify(partitionService, never()).recalculatePartitions(eq(currentInfo), any());
+
+        startNode(childData);
+
+        verify(partitionService, never()).recalculatePartitions(eq(currentInfo), any());
+
+        Thread.sleep(2000);
+
+        verify(partitionService, never()).recalculatePartitions(eq(currentInfo), any());
```
Comment (smatvienko-tb) [inline common/queue/src/test/java/org/thingsboard/server/queue/discovery/ZkDiscoveryServiceTest.java:118]: when you testing on `never` it is probably better to use `(any(), any())` matchers

File: common/queue/src/test/java/org/thingsboard/server/queue/discovery/ZkDiscoveryServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+        startNode(childData);
+
+        verify(partitionService, times(1)).recalculatePartitions(eq(currentInfo), eq(Collections.emptyList()));
+
+        verify(partitionService, times(1)).recalculatePartitions(eq(currentInfo), eq(List.of(childInfo)));
+
+        reset(partitionService);
+
+        //Start another node during restart
+        var anotherInfo = TransportProtos.ServiceInfo.newBuilder().setServiceId("anotherId").build();
```
Comment (smatvienko-tb) [inline common/queue/src/test/java/org/thingsboard/server/queue/discovery/ZkDiscoveryServiceTest.java:140]: real service name wanted

File: common/queue/src/test/java/org/thingsboard/server/queue/discovery/ZkDiscoveryServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+        assertEquals(1, zkDiscoveryService.delayedTasks.size());
+
+        startNode(anotherData);
+
+        assertTrue(zkDiscoveryService.delayedTasks.isEmpty());
+
+        verify(partitionService, times(1)).recalculatePartitions(eq(currentInfo), eq(List.of(anotherInfo)));
+        reset(partitionService);
+
+        Thread.sleep(2000);
```
Comment (smatvienko-tb) [inline common/queue/src/test/java/org/thingsboard/server/queue/discovery/ZkDiscoveryServiceTest.java:154]: :see_no_evil:

File: msa/vc-executor/src/main/resources/tb-vc-executor.yml [hunk 11]
```diff
@@ -41,6 +41,7 @@ zk:
   session_timeout_ms: "${ZOOKEEPER_SESSION_TIMEOUT_MS:3000}"
   # Name of the directory in zookeeper 'filesystem'
   zk_dir: "${ZOOKEEPER_NODES_DIR:/thingsboard}"
+  recalculate_delay: "${ZOOKEEPER_RECALCULATE_DELAY_MS:120000}"
```
Comment (smatvienko-tb) [inline msa/vc-executor/src/main/resources/tb-vc-executor.yml:44]: Two-minute downtime is quite noticeable for users. Can we do less by default?

File: common/queue/src/main/java/org/thingsboard/server/queue/discovery/ZkDiscoveryService.java [hunk 12]
```diff
@@ -290,8 +297,32 @@ public void childEvent(CuratorFramework curatorFramework, PathChildrenCacheEvent
...
+                } else {
+                    log.debug("[{}] Going to recalculate partitions due to adding new node [{}].",
+                            instance.getServiceId(), instance.getServiceTypesList());
+                    recalculatePartitions();
+                }
+                break;
             case CHILD_REMOVED:
-                recalculatePartitions();
+                ScheduledFuture<?> future = zkExecutorService.schedule(() -> {
+                    log.debug("[{}] Going to recalculate partitions due to removed node [{}]",
```
Comment (smatvienko-tb) [inline common/queue/src/main/java/org/thingsboard/server/queue/discovery/ZkDiscoveryService.java:318]: Please, improve logging. For the first time, the logs will be enabled on **debug** level to provide
Please, leave all spam logs on the TRACE level.
For now the debug logs is to spammy

```
2023-07-21 15:02:49,839 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Processing [CHILD_UPDATED] event for [tb-http-transport-1]
2023-07-21 15:02:49,920 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Ignoring event about current server PathChildrenCacheEvent{type=CHILD_UPDATED, data=ChildData{path='/thingsboard/nodes/0000001554', stat=55836430886,55836458616,....,57,0,55836430886
, data=[10, 19, 116, 98, 45, 115, 110, ... 71, 32, -128, -32, -12, -67, 30, 40, 20, 48, -128, -32, -2, -7, 127]}}
2023-07-21 15:02:56,910 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Processing [CHILD_UPDATED] event for [tb-mqtt-transport-2]
2023-07-21 15:02:57,487 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Processing [CHILD_UPDATED] event for [tb-core-1]
2023-07-21 15:03:02,889 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Processing [CHILD_UPDATED] event for [tb-ie-main-0]
2023-07-21 15:03:04,846 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Processing [CHILD_UPDATED] event for [tb-lwm2m-transport-0]
2023-07-21 15:03:05,442 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Processing [CHILD_UPDATED] event for [tb-rule-engine-0]
2023-07-21 15:03:05,445 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Processing [CHILD_UPDATED] event for [tb-vc-executor-1]
2023-07-21 15:03:06,703 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Processing [CHILD_UPDATED] event for [tb-vc-executor-0]
2023-07-21 15:03:14,163 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Processing [CHILD_UPDATED] event for [tb-ie-main-1]
2023-07-21 15:03:17,140 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Processing [CHILD_UPDATED] event for [tb-ie-main-2]
2023-07-21 15:03:17,940 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Processing [CHILD_UPDATED] event for [tb-core-2]
2023-07-21 15:03:19,812 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Processing [CHILD_UPDATED] event for [tb-mqtt-transport-0]
2023-07-21 15:03:23,395 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Processing [CHILD_UPDATED] event for [tb-rule-engine-1]
2023-07-21 15:03:29,300 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Processing [CHILD_UPDATED] event for [tb-core-3]
2023-07-21 15:03:30,644 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Processing [CHILD_UPDATED] event for [tb-core-0]
2023-07-21 15:03:32,277 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Processing [CHILD_UPDATED] event for [tb-coap-transport-0]
2023-07-21 15:03:33,715 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Processing [CHILD_UPDATED] event for [tb-http-transport-0]
2023-07-21 15:03:36,360 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Processing [CHILD_UPDATED] event for [tb-core-4]
2023-07-21 15:03:38,445 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Processing [CHILD_UPDATED] event for [tb-rule-engine-2]
2023-07-21 15:03:47,979 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Processing [CHILD_UPDATED] event for [tb-coap-transport-1]
2023-07-21 15:03:48,251 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Processing [CHILD_UPDATED] event for [tb-mqtt-transport-1]
2023-07-21 15:03:48,754 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Processing [CHILD_UPDATED] event for [tb-ie-other-0]

```

**Review by smatvienko-tb — APPROVED** [review]
I **believe** that this PR will work as expected.

It took some time to understand what is going on in real setup. 
I took examples from logs to find out will it work or not

I worried about that we have two point of concurrency:
1 - the map
2 - synchronized repartition with map.clear

Another concern is that after clear the futures still in scheduled pool and occupied  some resources

```
2023-07-21 14:59:39,859 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Processing [CHILD_REMOVED] event for [tb-mqtt-transport-2]
2023-07-21 14:59:55,927 [Curator-PathChildrenCache-0] DEBUG o.t.s.q.discovery.ZkDiscoveryService - Processing [CHILD_ADDED] event for [tb-mqtt-transport-2]
```

File: common/queue/src/test/java/org/thingsboard/server/queue/discovery/ZkDiscoveryServiceTest.java (hunk 7 above)
Comment (YevhenBondarenko) [reply common/queue/src/test/java/org/thingsboard/server/queue/discovery/ZkDiscoveryServiceTest.java:95]: this list is immutable
